src/archetype.rs

```diff
@@ -215,62 +251,98 @@ impl Archetypes {
 
+                // Look up if we have an archetype for the new component
+                // indices. If we do, insert it into the source archetype's
+                // `insert_components` map and return it. If we don't, create
+                // it.
                 match self
                     .by_components
                     .entry(new_components.into_boxed_slice().into())
                 {
                     Entry::Vacant(vacant_by_components) => {
+                        // Check for overflow of archetype indices.
                         assert!(next_arch_idx < u32::MAX as usize, "too many archetypes");
 
                         let arch_id = ArchetypeIdx(next_arch_idx as u32);
 
+                        // Construct the new archetype.
+                        // SAFETY:
+                        // - Component indices slice is in sorted order, as we created it by cloning
+                        //   a sorted slice and inserting a component index in the correct position
+                        //   that maintains the sorted order (see above).
+                        // - Similar rationale for validity of the indices: the source component
+                        //   indices slice was valid, and we only added an index that the caller of
+                        //   this function guaranteed to be valid.
+                        // - We only drop the component indices slice after dropping the archetype
+                        //   (or so we should; see FIXME note in `remove_component` above).
                         let mut new_arch = Archetype::new(
                             arch_id,
                             vacant_by_components.key().as_ref().into(),
                             components,
                         );
 
+                        // Insert a backlink to the source archetype.
                         new_arch
                             .remove_components
                             .insert(component_idx, src_arch_idx);
 
+                        // Register all event handlers for the new archetype.
                         for info in handlers.iter_mut() {
                             new_arch.register_handler(info);
                         }
 
+                        // Insert the archetype into the vacant entries and
+                        // our archetype list.
                         vacant_by_components.insert(arch_id);
-
                         vacant_insert_components.insert(arch_id);
-
                         self.archetypes.insert(new_arch);
 
                         arch_id
                     }
-                    Entry::Occupied(o) => *vacant_insert_components.insert(*o.get()),
+                    Entry::Occupied(entry) => {
+                        // The archetype already exists, insert it into the
+                        // map and return it.
+                        *vacant_insert_components.insert(*entry.get())
```

Comment:
Don't think this is comment is adding any value, it just restates how `insert` works